File: components/desktop/CswClient/trunk/src/CswSearchRequest.cs
/* See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * Esri Inc. licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;

namespace com.esri.gpt.csw
{
    /// <summary>
    /// CswSearchRequest class.
    /// </summary>
    /// <remarks>
    /// CswSearchRequest class is used to submit CSW search queries and to return CSW search results.
    /// Before submiting a request, you need to specify a catalog and provide search criteria.
    /// </remarks>
    public class CswSearchRequest
    {
        private CswCatalog _catalog;
        private CswSearchCriteria _criteria;
        private CswSearchResponse _response;
        private CswClient _cswClient;
        private string _mapServerUrl;  
        #region "Properties"
        /// <summary>
        /// Accessor methods
        /// </summary>
        public CswCatalog Catalog
        {
            get { return _catalog; }
            set { _catalog = value; }
        }

        public CswSearchCriteria Criteria
        {
            get { return _criteria; }
            set { _criteria = value; }
        }
        #endregion
        #region Constructor
        /// <summary>
        /// Constructor
        /// </summary>
        public CswSearchRequest() { }
        #endregion
        #region "Methods and Functions"
        /// <summary>
        /// Search CSW catalog using the provided criteria. Search result can be accessed by calling GetResponse(). 
        /// </summary>
        public void Search()
        {
            _response = new CswSearchResponse();
            StringBuilder sb = new StringBuilder();

            if (_criteria == null) {
                sb.AppendLine(DateTime.Now + " Criteria not specified.");
                throw new NullReferenceException("Criteria not specified."); }
            if (_catalog == null) {
                sb.AppendLine(DateTime.Now + " Catalog not specified.");
                throw new NullReferenceException("Catalog not specified."); }
            if (_catalog.URL == null || _catalog.URL.Length == 0)
            {
                sb.AppendLine(DateTime.Now + " Catalog URL not specified.");
                throw new NullReferenceException("Catalog URL not specified.");
            }
            if (_catalog.Profile == null)
            {
                sb.AppendLine(DateTime.Now + " Catalog profile not specified.");
                throw new NullReferenceException("Catalog profile not specified.");
            }

            CswProfile profile = _catalog.Profile;
            writeLogMessage("Csw profile used : " + profile.Name);
            // generate getRecords query
            string requestUrl = _catalog.Capabilities.GetRecords_PostURL;
            string requestQuery = profile.GenerateCSWGetRecordsRequest(_criteria);

         //   requestQuery = "<csw:GetCapabilities xmlns:csw=\"http://www.opengis.net/cat/csw/2.0.2\" request=\"GetCapabilities\" service=\"CSW\" version=\"2.0.2\"/>";

            requestQuery = requestQuery.Replace("utf-16", "utf-8");
            requestQuery = requestQuery.Replace("UTF-16", "UTF-8");

            // submit search query
            if (_cswClient == null) { _cswClient = new CswClient(); }

            string responseText;
            sb.AppendLine(DateTime.Now + " Sending CSW GetRecords request to endpoint : " + requestUrl);
            sb.AppendLine("Request Query : " + requestQuery);
            if(!_catalog.Capabilities.GetRecords_IsSoapEndPoint)
               responseText = _cswClient.SubmitHttpRequest("POST", requestUrl, requestQuery);
            else
               responseText = _cswClient.SubmitHttpRequest("SOAP", requestUrl, requestQuery);

            // parse out csw search records
            CswRecords records = new CswRecords();
            sb.AppendLine(DateTime.Now + " Response received : " + responseText);
            profile.ReadCSWGetRecordsResponse(responseText, records);
            sb.AppendLine(DateTime.Now + " Parsed GetRecords response.");

            // populate CSW response
            _response.ResponseXML = responseText;
            _response.Records = records;

            writeLogMessage(sb.ToString());
        }

        private void writeLogMessage(String logMessage){
                Utils.logger.writeLog(logMessage);
        }

        /// <summary>
        /// Retrieve metadata from CSW service by its ID 
        /// </summary>
        /// <param name="DocID">Metadata document ID</param>
        public bool GetMetadataByID(string DocID, bool bApplyTransform)
        {
            _response = new CswSearchResponse();
            StringBuilder sb = new StringBuilder();
            if (DocID == null || DocID == "") { throw new ArgumentNullException(); }
            if (_catalog == null) { throw new NullReferenceException("Catalog not specified."); }
            if (_catalog.Capabilities == null) { throw new NullReferenceException("Catalog capabilities not initialized."); }
            if (_catalog.Profile == null) { throw new NullReferenceException("Catalog profile not specified."); }

            if (_catalog.Capabilities.GetRecordByID_GetURL == null || _catalog.Capabilities.GetRecordByID_GetURL.Length == 0)
            {
                throw new NullReferenceException("GetRecordByID URL not specified for the catalog capabilities.");
            }

            CswProfile profile = _catalog.Profile;
            writeLogMessage(" Csw profile used : " + profile.Name);
            // generate request url
            string getRecordByIDBaseUrl = _catalog.Capabilities.GetRecordByID_GetURL;
            string requestUrl = profile.GenerateCSWGetMetadataByIDRequestURL(getRecordByIDBaseUrl, DocID);

            sb.AppendLine(DateTime.Now + " GetRecordsById request URL : " + requestUrl);
            if (_cswClient == null) { _cswClient = new CswClient(); }
            string responseText = _cswClient.SubmitHttpRequest("GET", requestUrl, "");
            _response.ResponseXML = responseText;
            sb.AppendLine(DateTime.Now + " GetRecordsById response xml : " + responseText);
            CswRecord record = new CswRecord(DocID);
            bool isTransformed = false;
            if (bApplyTransform)
            {
                isTransformed = profile.TransformCSWGetMetadataByIDResponse(responseText, record);
                if(isTransformed)
                    sb.AppendLine(DateTime.Now + " Transformed xml : " + record.FullMetadata);
            }
            else
            {
                record.FullMetadata = responseText;
            }

            /*if (!isTransformed)
            {
                XmlDocument responseXml = new XmlDocument();
                try { responseXml.LoadXml(responseText); }
                catch (XmlException xmlEx)
                {
                    throw new XmlException("Error occurred \r\n" + xmlEx.Message);
                }
                record.FullMetadata = responseXml.FirstChild.InnerText ;
            }*/

            // add record to the response
            CswRecords records = new CswRecords();
            if (record != null) { records.Add(record.ID, record); }
            _response.Records = records;

            _mapServerUrl = record.MapServerURL;

            if (_mapServerUrl != null)
                sb.AppendLine(DateTime.Now + " Map Server Url : " + _mapServerUrl);

            writeLogMessage(sb.ToString());

            return isTransformed;

        }
        
        /// <summary>
        /// Get Add to map information
        /// </summary>
        /// <param name="DocID">document identifier</param>
        public void GetAddToMapInfoByID(string DocID)
        {

            _response = new CswSearchResponse();
            StringBuilder sb = new StringBuilder();

            if (DocID == null || DocID == "") { throw new ArgumentNullException(); }
            if (_catalog == null) { throw new NullReferenceException("Catalog not specified."); }
            if (_catalog.Capabilities == null) { throw new NullReferenceException("Catalog capabilities not initialized."); }
            if (_catalog.Profile == null) { throw new NullReferenceException("Catalog profile not specified."); }

            if (_catalog.Capabilities.GetRecordByID_GetURL == null || _catalog.Capabilities.GetRecordByID_GetURL.Length == 0)
            {
                sb.AppendLine(DateTime.Now + " GetRecordByID URL not specified for the catalog capabilities.");
                throw new NullReferenceException("GetRecordByID URL not specified for the catalog capabilities.");
            }

            CswProfile profile = _catalog.Profile;

            // generate request url
            string getRecordByIDBaseUrl = _catalog.Capabilities.GetRecordByID_GetURL;
            string requestUrl = profile.GenerateCSWGetMetadataByIDRequestURL(getRecordByIDBaseUrl, DocID);

            sb.AppendLine(DateTime.Now + " GetRecordsById request URL : " + requestUrl);
            if (_cswClient == null) { _cswClient = new CswClient(); }
            string responseText = _cswClient.SubmitHttpRequest("GET", requestUrl, "");
            _response.ResponseXML = responseText;

            sb.AppendLine(DateTime.Now + " GetRecordsById response xml : " + responseText);
            CswRecord record = new CswRecord(DocID);

            profile.ReadCSWGetMetadataByIDResponse(responseText, record);

                if (record.MetadataResourceURL != null && record.MetadataResourceURL.Equals("") && profile.Name.Equals("terra catalog CSW 2.0.2 AP ISO"))
                {
                    record.FullMetadata = responseText;
                }

                if (record == null) { throw new NullReferenceException("Record not populated."); }

                // check if full metadata or resourceURL has been returned
                bool hasFullMetadata = !(record.FullMetadata == null || record.FullMetadata == "");
                bool hasResourceUrl = !(record.MetadataResourceURL == null || record.MetadataResourceURL == "");
                if (!hasFullMetadata && !hasResourceUrl)
                {
                   // throw new InvalidOperationException("Neither full metadata nor metadata resource URL was found for the CSW record.");
                }
                else if (!hasFullMetadata && record.MetadataResourceURL != null)
                {
                    // need to load metadata from resource URL
                    responseText = _cswClient.SubmitHttpRequest("GET", record.MetadataResourceURL, "", "", "");
                    record.FullMetadata = responseText;
                }

                // add record to the response
                CswRecords records = new CswRecords();
                if (record != null) { records.Add(record.ID, record); }
                _response.Records = records;

                _mapServerUrl = record.MapServerURL;

                if (_mapServerUrl != null)
                    sb.AppendLine(DateTime.Now + " Map Server Url : " + _mapServerUrl);

                writeLogMessage(sb.ToString());
    }



        /// <summary>
        /// Get the CSW search response of a CSW search request 
        /// </summary>
        /// <remarks>
        /// Get the CSW search response of a CSW search request 
        /// </remarks>
        /// <returns>a CswSearchResponse object.</returns>
        public CswSearchResponse GetResponse()
        {
            return _response;
        }
        /// <summary>
        /// Map server url
        /// </summary>
        /// <returns></returns>
        public string GetMapServerUrl()
        {
            return _mapServerUrl;
        }
        #endregion
    }
}

File: geoportal/src/com/esri/gpt/server/csw/client/CswSearchRequest.java
/* See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * Esri Inc. licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.esri.gpt.server.csw.client;

import java.io.*;
import java.util.Iterator;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.TransformerException;
import javax.xml.xpath.XPathExpressionException;
import org.xml.sax.SAXException;

/**
 * CswSearchRequest class.
 * 
 * CswSearchRequest class is used to submit CSW search queries and to
 * return CSW search results. Before submiting a request, you need to specify a
 * catalog and provide search criteria.
 */
public class CswSearchRequest {

private static final Logger LOG = 
  Logger.getLogger(CswSearchRequest.class.getCanonicalName());

private CswCatalog        catalog;
private CswSearchCriteria criteria;
private CswClient         cswClient;
private CswSearchResponse response;


public CswSearchRequest() {
  this(null, null);
}

public CswSearchRequest(CswCatalog catalog, String searchText) {
  this.catalog = catalog;
  // Initialize the neccessary objects
  // create search criteria
  this.criteria = new CswSearchCriteria();
  this.criteria.setSearchText(searchText);
  // create csw client
  this.cswClient = new CswClient();
  this.response  = new CswSearchResponse();
}

/**
 * Accessor methods
 */
public CswCatalog getCatalog() {
  return this.catalog;
}

public void setCatalog(CswCatalog catalog) {
  this.catalog = catalog;
  this.cswClient.setBatchHttpClient(catalog.getBatchHttpClient());
}

public CswSearchCriteria getCriteria() {
  return this.criteria;
}

public void setCriteria(CswSearchCriteria criteria) {
  this.criteria = criteria;
}

public CswClient getCswClient() {
  return this.cswClient;
}

public void setCswClient(CswClient cswClient) {
  this.cswClient = cswClient;
}

public CswSearchResponse getCswSearchResponse() {
  return this.response;
}

public void setCswSearchResponse(CswSearchResponse response) {
  this.response = response;
}

/**
 * Retrieve metadata from CSW service by its ID
 * 
 * @param DocID    Metadata document ID
 */
public void getMetadataByID(String DocID) throws NullReferenceException,
    IOException, InvalidOperationException, TransformerException {
  if (DocID == null || DocID.length() == 0) {
    throw new NullReferenceException("No DocID specified");
  }
  if (catalog == null) {
    throw new NullReferenceException("Catalog not specified.");
  }
  if (catalog.getCapabilities() == null) {
    throw new NullReferenceException("Catalog capabilities not initialized.");
  }
  if (catalog.getProfile() == null) {
    throw new NullReferenceException("Catalog profile not specified.");
  }

  if (catalog.getCapabilities().get_getRecordByIDGetURL() == null
      || catalog.getCapabilities().get_getRecordByIDGetURL().length() == 0) {
    throw new NullReferenceException(
        "GetRecordByID URL not specified for the catalog capabilities.");
  }

  CswProfile profile = catalog.getProfile();

  // generate request url
  String getRecordByIDBaseUrl = catalog.getCapabilities()
      .get_getRecordByIDGetURL();

  CswRecord record = this.getRecordById(getRecordByIDBaseUrl, DocID, profile);
  
 
  // add record to the response
  CswRecords records = new CswRecords();
  if (record != null) {
    records.add(record);
  }
  response.setRecords(records);

}

/**
 * Gets the record by id.
 * 
 * @param requestURL the request URL (CSW service url)
 * @param DocID the doc ID (GUID or UUID)
 * @param profile the profile
 * 
 * @return the record by id
 * 
 * @throws IOException Signals that an I/O exception has occurred.
 * @throws TransformerException the transformer exception
 * @throws NullReferenceException the null reference exception
 * @throws InvalidOperationException the invalid operation exception
 */
@SuppressWarnings("unchecked")
public CswRecord getRecordById(String requestURL, String DocID,
    CswProfile profile) throws IOException, TransformerException,
    NullReferenceException, InvalidOperationException {

  String requestUrl = profile.generateCSWGetMetadataByIDRequestURL(requestURL,
      DocID);

  if (cswClient == null) {
    cswClient = new CswClient();
    cswClient.setBatchHttpClient(catalog.getBatchHttpClient());
  }

  BufferedInputStream bStream = null;
  InputStream istIntermidiateDoc = null;
  InputStream istRealDoc = null;
  CswRecord record = null;

  try {
    istIntermidiateDoc = 
      cswClient.submitHttpRequest("GET", requestUrl,
      "");

    
    String responseStr = Utils.getInputString2(istIntermidiateDoc);
    
    LOG.log(Level.FINER, "Get Record By Id intermidiate XML = {0}", responseStr);
    response.setResponseXML(Utils.chkStr(responseStr));

    record = new CswRecord();
    CswRecords recordList = null;
    try {
      LOG.finer("GetRecordById: Making csw record object using g" +
      		"etRecordsResponse operation");
      // making cswObject by going through getrecord response xslt
      CswResult results = new CswResult();
      profile.readGetRecordsResponse(responseStr, results);
      recordList = results.getRecords();
      Iterator iter = recordList.iterator();
      if(iter.hasNext()) {
        Object obj = iter.next();
        if(obj instanceof CswRecord ) {
          record = (CswRecord) obj;
        }
      } else  {
        LOG.log(Level.WARNING, 
            "Could not get csw metadata of metadata document");
      }
    } catch (ParserConfigurationException e) {
      LOG.log(Level.INFO, 
        "Could not get csw metadata of metadata document (maybe this csw does " 
        + "not have csw metadata on getRecord by id. " + "{0}", e.getMessage());
    } catch (SAXException e) {
      LOG.log(Level.INFO, 
        "Could not get csw metadata of metadata document (maybe this csw does " 
        + "not have csw metadata on getRecord by id. " + "{0}", e.getMessage());
    } catch (XPathExpressionException e) {
      LOG.log(Level.INFO, 
        "Could not get csw metadata of metadata document (maybe this csw does " 
        + "not have csw metadata on getRecord by id. " + "{0}", e.getMessage());
    }
    record.setId(DocID);
    LOG.finer("GetRecordByID: Transforming intermidiate xml to populate xml " +
    		"into csw Record");
    profile.readCSWGetMetadataByIDResponse(getCswClient(),responseStr, record);
    if (record == null) {
      throw new NullReferenceException("Record not populated.");
    }
    

    // check if full metadata or resourceURL has been returned
    boolean hasFullMetadata = !(record.getFullMetadata() == null || record
        .getFullMetadata().equals(""));
    boolean hasResourceUrl = !(record.getMetadataResourceURL() == null || record
        .getMetadataResourceURL().equals(""));
    if(!hasResourceUrl && requestUrl != null ) {
      record.setMetadataResourceURL(requestUrl);
      hasResourceUrl = true;
    } 
    
    if (!hasFullMetadata && !hasResourceUrl) {
      throw new InvalidOperationException("Neither full metadata nor metadata"
          + " resource URL was found for the CSW record.");
    } 
  }
  finally {
    Utils.close(istIntermidiateDoc);
    Utils.close(bStream);
  }
  return record;

}

/**
 * Gets the record by id.
 * 
 * @param requestURL the request URL (CSW service url)
 * @param DocID the doc ID (GUID or UUID)
 * @param profile the profile
 * 
 * @return the record by id
 * 
 * @throws IOException Signals that an I/O exception has occurred.
 * @throws TransformerException the transformer exception
 * @throws NullReferenceException the null reference exception
 * @throws InvalidOperationException the invalid operation exception
 */
@SuppressWarnings("unchecked")
public CswRecord getRecordById(String requestURL, String DocID,
    CswProfile profile, String username, String password) 
    throws IOException, TransformerException,
    NullReferenceException, InvalidOperationException {

  String requestUrl = profile.generateCSWGetMetadataByIDRequestURL(requestURL,
      DocID);

  if (cswClient == null) {
    cswClient = new CswClient();
    cswClient.setBatchHttpClient(catalog.getBatchHttpClient());
  }

  BufferedInputStream bStream = null;
  InputStream istIntermidiateDoc = null;
  InputStream istRealDoc = null;
  CswRecord record = null;

  try {
    istIntermidiateDoc = 
      cswClient.submitHttpRequest("GET", requestUrl,
      "", username, password);

    
    String responseStr = Utils.getInputString2(istIntermidiateDoc);
    
    LOG.log(Level.FINER, "Get Record By Id intermidiate XML = {0}", responseStr);
    response.setResponseXML(Utils.chkStr(responseStr));

    record = new CswRecord();
    CswRecords recordList = new CswRecords();
    try {
      LOG.finer("GetRecordById: Making csw record object using g" +
          "etRecordsResponse operation");
      // making cswObject by going through getrecord response xslt
      profile.readCSWGetRecordsResponse(responseStr, recordList);
      Iterator iter = recordList.iterator();
      if(iter.hasNext()) {
        Object obj = iter.next();
        if(obj instanceof CswRecord ) {
          record = (CswRecord) obj;
        }
      } else  {
        LOG.log(Level.WARNING, 
            "Could not get csw metadata of metadata document");
      }
    } catch (ParserConfigurationException e) {
      LOG.log(Level.INFO, 
        "Could not get csw metadata of metadata document (maybe this csw does " 
        + "not have csw metadata on getRecord by id. " + "{0}", e.getMessage());
    } catch (SAXException e) {
      LOG.log(Level.INFO, 
        "Could not get csw metadata of metadata document (maybe this csw does " 
        + "not have csw metadata on getRecord by id. " + "{0}", e.getMessage());
    } catch (XPathExpressionException e) {
      LOG.log(Level.INFO, 
        "Could not get csw metadata of metadata document (maybe this csw does " 
        + "not have csw metadata on getRecord by id. " + "{0}", e.getMessage());
    }
    record.setId(DocID);
    LOG.finer("GetRecordByID: Transforming intermidiate xml to populate xml " +
        "into csw Record");
    profile.readCSWGetMetadataByIDResponse(getCswClient(), responseStr, record);
    if (record == null) {
      throw new NullReferenceException("Record not populated.");
    }
    

    // check if full metadata or resourceURL has been returned
    boolean hasFullMetadata = !(record.getFullMetadata() == null || record
        .getFullMetadata() == "");
    boolean hasResourceUrl = !(record.getMetadataResourceURL() == null || record
        .getMetadataResourceURL() == "");
    
    if(!hasResourceUrl && requestUrl != null ) {
      record.setMetadataResourceURL(requestUrl);
      hasResourceUrl = true;
    }

    if (!hasFullMetadata && !hasResourceUrl) {
      throw new InvalidOperationException("Neither full metadata nor metadata"
          + " resource URL was found for the CSW record.");
    } else if (hasResourceUrl) {
      // need to load metadata from resource URL
      istRealDoc = cswClient.submitHttpRequest("GET", Utils.chkStr(record
          .getMetadataResourceURL()), "", username, password);

      responseStr = Utils.getInputString2(istRealDoc);

      record.setFullMetadata(responseStr);
    }
  }
  finally {
    Utils.close(istIntermidiateDoc);
    Utils.close(bStream);
  }
  return record;

}



/**
 * Get the CSW search response of a CSW search request
 * 
 * Get the CSW search response of a CSW search request
 * @return a CswSearchResponse object.
 */
public CswSearchResponse getResponse() {
  return this.response;
}

/**
 * Search CSW catalog using the provided criteria. Search result can be accessed
 * by calling GetResponse().
 * @throws IOException 
 * @throws SAXException 
 * @throws ParserConfigurationException 
 * @throws XPathExpressionException 
 * @throws TransformerException 
 */
public void search() throws NullReferenceException, XPathExpressionException,
    ParserConfigurationException, SAXException, IOException,
    TransformerException {
  // Check the necessary info. to search csw
  if (catalog == null) {
    throw new NullReferenceException("No catalog specified");
  }
  if (criteria == null) {
    throw new NullReferenceException("No specified Criteria");
  }
  if (catalog.getUrl() == null || catalog.getUrl().length() == 0) {
    throw new NullReferenceException("No specified url");
  }
  if (catalog.getProfile() == null) {
    throw new NullReferenceException("No specified profile");
  }

  // Generate getRecords query
  CswProfile profile = catalog.getProfile();

  catalog.connect();
  CswCatalogCapabilities capabilities = catalog.getCapabilities();

  String requestUrl = capabilities.get_getRecordsPostURL();
  String requestQuery = profile.generateCSWGetRecordsRequest(criteria);

  
  // Submit search query and get response as an InputStream object
  InputStream responseStream = cswClient.submitHttpRequest("POST", requestUrl,
      requestQuery);
 
  BufferedInputStream bIStream = new BufferedInputStream(responseStream);
 
  try {
//    String responseStr = Utils.getInputString(bIStream);
    String responseStr = readCharacters(bIStream, "UTF-8");
    // Transform input xml to output xml with Profile
    CswRecords records = new CswRecords();

    profile.readCSWGetRecordsResponse(responseStr, records);

    response.set_requestStr(requestQuery);
    response.setRecords(records);
    response.setResponseXML(responseStr);
  } finally {
    Utils.close(bIStream);
    Utils.close(responseStream);
  }

}

/**
 * Fully reads the characters from an input stream.
 * @param stream the input stream
 * @param charset the encoding of the input stream
 * @return the characters read
 * @throws IOException if an exception occurs
 */
private String readCharacters(InputStream stream, String charset)
  throws IOException {
  StringBuilder sb = new StringBuilder();
  BufferedReader br = null;
  InputStreamReader ir = null;
  try {
    if ((charset == null) || (charset.trim().length() == 0)) charset = "UTF-8";
    char cbuf[] = new char[2048];
    int n = 0;
    int nLen = cbuf.length;
    ir = new InputStreamReader(stream,charset);
    br = new BufferedReader(ir);
    while ((n = br.read(cbuf,0,nLen)) > 0) sb.append(cbuf,0,n);
  } finally {
    try {if (br != null) br.close();} catch (Exception ef) {}
    try {if (ir != null) ir.close();} catch (Exception ef) {}
  }
  return sb.toString();
}

}